nemo/collections/tts/models/audio_codec.py

```diff
@@ -473,10 +477,26 @@ def validation_step(self, batch, batch_idx):
         }
         self.log_dict(metrics, on_epoch=True, sync_dist=True)
 
+    @staticmethod
+    def get_dataset(cfg):
+        if 'is_sharded' in cfg.dataset:
+            is_sharded = cfg.dataset.is_sharded
+            del cfg.dataset.is_sharded
+        else:
+            is_sharded = False
+
+        if is_sharded:
+            cfg.dataset._target_ = 'nemo.collections.tts.data.vocoder_dataset.TarredVocoderDataset'
+            dataset = instantiate(cfg.dataset)
+            sampler = None
+        else:
+            dataset = instantiate(cfg.dataset)
+            sampler = dataset.get_sampler(cfg.dataloader_params.batch_size)
+        return dataset, sampler
```

Comment:
Good point i will create a simpler case here, but I created this function to set the target so as to use existing config files and provide toggle option for sharding.

Reply:
I am against creating one new config for each feature.

Reply:
Do you mean you are against creating separate config files for new features in general (eg. sampling rate, new architectures)? Or specifically for needing separate config for tarred vs non-tarred datasets?

The former of creating lots of different overlapping configs is expected across NeMo. The latter can be addressed in several different ways, if needed.

Reply:
We don;t want to create multiple configs with same model architecture but different setting (like tarred/sampling_rate etc).

Reply:
For the same model architecture with different configs are main options I have heard before are:

1. Create different example configs.
2. Have one example config and provide web documentation and tutorials clearly indicating to users how to manually change fields for different valid configs (eg. sample rate, language).

From past discussions around this, generally (1) was preferable, especially for changes like sample rate which are relatively complicated and error prone if you attempt to update all needed parameters manually.  A combination of (1) and (2) is needed if you have too many configs to support, such as when having to support multiple sample rates and languages at the same time.

Reply:
In this particular case, there are 2 problems with the existing approach.

- The model class takes the data loader class/config as input, however the constructor requires passing information `global_rank` and `world_size` which are known only to the model class and trainer.
- Different data loaders require significantly different inputs to instantiate (and this will diverge more as we add additional sampling strategies).

Perhaps the cleanest way to address these would be to change it so the model is expected to receive a subset of dataset parameters, and then add a module which maps the parameters to the logic to create each dataset.

For example:

**In .yaml**
```
train_ds:
  dataset_type: "vocoder_tarred"
  dataset_args:
    batch_size: ${batch_size}
    ... 
```

**In tts_dataset_utils.py**

```
def create_dataset(dataset_type, dataset_args, global_rank, world_size, ...):
    if dataset_type == "vocoder":
        return VocoderDataset(*dataset_args)
    elif dataset_type == "vocoder_tarred":
        return TarrredVocoderDataset(*dataset_args, global_rank=global_rank, world_size=world_size)
     ...
```

Reply:
Good catch, missed it. Updated to use those as well.

Reply:
for logging, testing, validation we rely only on non tarred datasets.